File: Residential_Controller.rb
class Column
    attr_accessor :floorAmount, :elevatorAmount, :elevatorList
    attr_reader :floorButtonList, :elevatorList, :elevatorAmount, :floorAmount
    def initialize(floorAmount, elevatorAmount)
        @floorAmount= floorAmount
        @elevatorAmount =elevatorAmount
        @elevatorList = Array.new
        @floorButtonList = Array.new 
       
        self.createElevatorList() 
        self.createFloorButtonList() 
    end
    def elevatorList
     @elevatorList
    end

    def createElevatorList()
        for i in 1..@elevatorAmount        
            elevator = Elevator.new(i, "idle", 1, "up", @floorAmount) 
            @elevatorList.push(elevator) 
        end
    end
    def createFloorButtonList ()
        button = FloorButton.new(1, "up") 
        @floorButtonList.push(button)
        for i in 2..(@floorAmount-1)        
            button = FloorButton.new(i, "up") 
            @floorButtonList.push(button) 
            button = FloorButton.new(i, "down") 
            @floorButtonList.push(button)
        end
        button = FloorButton.new(@floorAmount, "up") 
        @floorButtonList.push(button)
    end
   
    def requestElevator(requestedFloor, direction) 
        elevator = self.bestElevator(requestedFloor, direction)
      
        elevator.requestList.push(requestedFloor)
        print("----------------------------------\n")
        puts("\nRequest elevator from floor: "+ requestedFloor.to_s+ "  direction: "+ direction )
        puts("Selected elevator id: "+elevator.id.to_s+
        "  position: "+ elevator.position.to_s)
        print("----------------------------------\n")
        elevator.manageRequestList()
        elevator.moveElevator(requestedFloor)
        # puts("Elevator = ", elevator.id, "position: ", elevator.position,	"Status: ", elevator.status)
        return elevator
    end
    def bestElevator(requestedFloor, direction)
        bestfit = nil
        minDistance = 11
        nearestIdle = nil
        for elevator in (self.elevatorList)
            puts("elevator ID: "+ (elevator.id).to_s+ "  position:  "+ (elevator.position).to_s+ "  direction:  "+ elevator.direction+ "Status:  "+ elevator.status)
            distance = (requestedFloor - elevator.position).abs
            if (requestedFloor == elevator.position and (elevator.status == "stopped" or elevator.status == "idle"))
                bestfit = elevator
            elsif (minDistance > distance and elevator.direction == direction)
                minDistance = distance
                bestfit = elevator
            end
        end
        idleDistance = 11
        for elevator in (self.elevatorList)
            distance = (requestedFloor - elevator.position).abs
            if (elevator.status == "idle" and idleDistance >= distance)
                idleDistance = distance
                nearestIdle = elevator
            end
        end

        if bestfit != nil
            return bestfit
        else
            return nearestIdle
        end
    
    end
    def requestFloor(elevator, requestedFloor)
        puts("\nlanding request to floor: [ "+ requestedFloor.to_s+ "]")
        elevator.requestList.push(requestedFloor)
        elevator.manageRequestList()
        elevator.moveElevator(requestedFloor) 
           
    end
end
class Elevator
    attr_reader :id, :status, :position, :direction, :panelButtonList, :requestList
    attr_accessor :id, :status, :position, :direction, :panelButtonList, :requestList
    def initialize(id, status, position, direction, floorAmount)
        @id = id
        @status = status
        @position = position
        @direction = direction
        @requestList = Array.new
        @panelButtonList = Array.new 
        
        def status
     @status
    end
        for i in 1..9        
            button = PanelButton.new(i) 
            @panelButtonList.push(button) 
        end
    end
    def manageRequestList()
            if (self.direction == "up")
                self.requestList.sort()
            elsif (self.direction == "down")
                self.requestList.sort()
                self.requestList.reverse() 
            end
    end
    def moveUp(requestedFloor)
            while (self.position != requestedFloor)
                self.position += 1
                puts(">> Elevator is moving to : [ "+ self.position.to_s+" ]")
            end
    end
    def moveDown(requestedFloor)
        while (self.position != requestedFloor)
            self.position -= 1
            puts(">> Elevator is moving to : [ "+ self.position.to_s+" ]")
        end
    end

    def moveElevator(requestedFloor)
        while (self.requestList.length() > 0)
            if (requestedFloor == self.position)
                self.status = "stopped"
                puts("Elevator: [ "+ self.id.to_s+ " ] sent to [ " +requestedFloor.to_s+ " ]")
                self.openDoor()
                self.status = "moving"
                self.requestList.pop()
                elsif (requestedFloor < self.position)
                self.status = "moving"
                self.direction = "down"
                puts("Elevator: ["+ self.id.to_s+ " ] Status: ["+
                      self.status+ " ] direction: [ "+ self.direction+" ]")
                self.moveDown(requestedFloor)
                self.status = "stopped"
                puts("Elevator: ["+ self.id.to_s+ " ] Status: ["+
                    self.status+ " ] direction: [ "+ self.direction+" ]")
                self.openDoor()
                self.requestList.pop()
                elsif (requestedFloor > self.position)
                self.status = "moving"
                self.direction = "up"
                puts("Elevator: [ "+ self.id.to_s+ " ] Status: [ "+
                      self.status+ " ] direction: ["+ self.direction+ "]")
                self.moveUp(requestedFloor)
                self.status = "stopped"
                puts("Elevator: [ "+ self.id.to_s+ " ] Status: [ "+
                    self.status)
                self.openDoor()
                self.requestList.pop()
            end
        end

        if (self.requestList == 0)
            self.status = "idle"
            puts("Elevator: "+  self.status)
        end
    end
    def openDoor()
        print("[ open Door ]\n")
        self.closeDoor()
    end

    def closeDoor()
        print("[close Door ]\n")
    end
end
class FloorButton
    def initialize(numFloor, direction)
        @numFloor = numFloor
        @direction = direction
    end
end


class PanelButton
    def initialize(destination)
        @destination = destination
    end
end


# testing 

print("\n-----------     [ Scenario 1 ] --------------   \n ")
print(" 1. Someone is on floor 3 and wants to go to the 7th floor\n")
print(" \n ")

column1 = Column.new(10, 2)

column1.elevatorList[0].status = "idle"
column1.elevatorList[0].position = 2
column1.elevatorList[0].direction = "none"
column1.elevatorList[1].status = "idle"
column1.elevatorList[1].position = 6
column1.elevatorList[1].direction = "none"
elevator = column1.requestElevator(3, "up")
column1.requestFloor(elevator, 7)
puts("\n")
print("\n-----------     [ Scenario 2 ] --------------   \n ")
print(" 2.1. Someone is on the 1st floor and requests the 6th floor. \n\n")
print(" \n ")

column1.elevatorList[0].status = "idle"
column1.elevatorList[0].position = 10
column1.elevatorList[0].direction = "none"
column1.elevatorList[1].status = "idle"
column1.elevatorList[1].position = 3
column1.elevatorList[1].direction = "none"

elevator = column1.requestElevator(1, "up")
column1.requestFloor(elevator, 6)
print("  \n ")
print("\n2.2. someone else is on the 3rd floor and requests the 5th floor. \n")
print(" \n ")
elevator = column1.requestElevator(3, "up")
column1.requestFloor(elevator, 5)
print(" \n ")
print("\n 2.3. a third person is at floor 9 and wants to go down to the 2nd floor.  \n")
print(" \n ")
elevator = column1.requestElevator(9, "down")
column1.requestFloor(elevator, 2)
puts("\n")
print("\n-----------     [ Scenario 3 ] --------------   \n ")
print(" 3.1. Someone is on floor 3 and requests the 2nd floor.  \n")
print("  \n ")

column1.elevatorList[0].status = "idle"
column1.elevatorList[0].position = 10
column1.elevatorList[0].direction = "none"
column1.elevatorList[1].status = "moving"
column1.elevatorList[1].position = 3
column1.elevatorList[1].direction = "up"

elevator = column1.requestElevator(3, "down")
column1.requestFloor(elevator, 2)
print("\n**********  \n ")
print("\n3.2. someone else is on the 10th floor and wants to go to the 3rd.\n ")
print("  \n ")


column1.elevatorList[0].status = "idle"
column1.elevatorList[0].position = 2
column1.elevatorList[0].direction = "none"
column1.elevatorList[1].status = "idle"
column1.elevatorList[1].position = 6
column1.elevatorList[1].direction = "none"

elevator = column1.requestElevator(10, "down")
column1.requestFloor(elevator, 3)

File: Residential_Controller.py
# --------------------------------------
# Residential controller python version
# --------------------------------------

class Column:
    def __init__(self, floorAmount, elevatorAmount):
        self.floorAmount = floorAmount
        self.elevatorAmount = elevatorAmount
        self.elevatorList = []
        for i in range(elevatorAmount):
            elevator = Elevator(i + 1, "idle", 1, "up", floorAmount)
            self.elevatorList.append(elevator)
        self.floorButtonList = []
        for i in range(floorAmount):
            if i == 1:
                self.floorButtonList.append(
                    floorButton(i, "up"))  # first floor
            self.floorButtonList.append(floorButton(i, "up"))
            self.floorButtonList.append(floorButton(i, "down"))
        self.floorButtonList.append(floorButton(
            floorAmount, "down"))  # last floor

    def requestElevator(self, requestedFloor, direction):
        elevator = self.bestElevator(requestedFloor, direction)
        elevator.requestList.append(requestedFloor)

        print(bcolors.OKGREEN+"\nRequest elevator from floor: "+str(requestedFloor) +
              "  direction: " + direction + bcolors.ENDC)
        print(bcolors.UNDERLINE+"elevator selected id: " + bcolors.ENDC, elevator.id,
              "position: ", elevator.position)
        elevator.manageRequestList()
        elevator.moveElevator(requestedFloor)
        print("Elevator = ", elevator.id, "position: ",
              elevator.position,	"Status: ", elevator.status)
        return elevator

    def bestElevator(self, requestedFloor, direction):
        bestfit = None
        minDistance = 11
        nearestIdle = None
        for elevator in (self.elevatorList):
            print("elevator ID [ "+bcolors.WARNING + str(elevator.id) + bcolors.ENDC+" ]  position: [ "+bcolors.WARNING + str(elevator.position)+bcolors.ENDC +
                  " ] direction: [ "+bcolors.WARNING + str(elevator.direction)+bcolors.ENDC+"] Status: [ "+bcolors.WARNING + str(elevator.status) + bcolors.ENDC + "]")
            distance = abs(requestedFloor - elevator.position)
            if (requestedFloor == elevator.position and (elevator.status == "stopped" or elevator.status == "idle")):
                bestfit = elevator
            elif (minDistance > distance and elevator.direction == direction):
                minDistance = distance
                bestfit = elevator
        idleDistance = 11
        for elevator in (self.elevatorList):
            distance = abs(requestedFloor - elevator.position)
            if (elevator.status == "idle" and idleDistance >= distance):
                idleDistance = distance
                nearestIdle = elevator

        if bestfit != None:
            return bestfit
        else:
            return nearestIdle

    def requestFloor(self, elevator, requestedFloor):
        print(bcolors.OKGREEN+"\nlanding request to floor: " +
              str(requestedFloor)+bcolors.ENDC)
        elevator.requestList.append(requestedFloor)
        elevator.manageRequestList()
        elevator.moveElevator(requestedFloor)


class Elevator:
    def __init__(self, id, status, position, direction, floorAmount):
        self.id = id
        self.status = status
        self.position = position
        self.direction = direction
        self.requestList = []
        self.panelButtonList = []
        for i in range(floorAmount):
            button = panelButton(i)
            self.panelButtonList.append(button)

    def manageRequestList(self):
        if self.direction == "up":
            self.requestList.sort()
        elif self.direction == "down":
            self.requestList.sort()
            self.requestList.reverse()

    def moveUp(self, requestedFloor):
        while (self.position != requestedFloor):
            self.position += 1
            print(">> Elevator is moving to :  " +
                  bcolors.OKBLUE + str(self.position)+bcolors.ENDC)

    def moveDown(self, requestedFloor):
        while (self.position != requestedFloor):
            self.position -= 1
            print(">> Elevator is moving to :  " +
                  bcolors.OKBLUE + str(self.position)+bcolors.ENDC)

    def moveElevator(self, requestedFloor):
        while (len(self.requestList) > 0):
            if (requestedFloor == self.position):
                self.status = "stopped"
                print("Elevator: ", self.id, " Status: ",
                      self.status, " direction: ", self.direction)
                self.openDoor()
                self.status = "moving"
                self.requestList.pop()
            elif (requestedFloor < self.position):
                self.status = "moving"
                self.direction = "down"
                print("Elevator: ", self.id, " Status: ",
                      self.status, " direction: ", self.direction)
                self.moveDown(requestedFloor)
                self.status = "stopped"
                print("Elevator: ", self.id, " Status: ",
                      self.status, " direction: ", self.direction)
                self.openDoor()
                self.requestList.pop()
            elif (requestedFloor > self.position):
                self.status = "moving"
                self.direction = "up"
                print("Elevator: ", self.id, " Status: ",
                      self.status, " direction: ", self.direction)
                self.moveUp(requestedFloor)
                self.status = "stopped"
                print("Elevator: ", self.id, " Status: ",
                      self.status)
                self.openDoor()
                self.requestList.pop()

        if self.requestList == 0:
            self.status = "idle"
            print("Elevator: ",  self.status)

    def openDoor(self):
        print(bcolors.WARNING+"[-- open Door -- ]"+bcolors.ENDC)
        self.closeDoor()

    def closeDoor(self):
        print(bcolors.WARNING+"[-- close Door-- ]"+bcolors.ENDC)


class floorButton:
    def __init__(self, numFloor, direction):
        self.numFloor = numFloor
        self.direction = direction


class panelButton:
    def __init__(self, destination):
        self.destination = destination

# ---------------------------------------
# ---------  [ Testing Section ] -------


class bcolors:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'


print(bcolors.HEADER + bcolors.UNDERLINE + bcolors.BOLD +
      "     [ Scenario 1 ]     "+bcolors.ENDC)
print(bcolors.HEADER +
      " 1. Someone is on floor 3 and wants to go to the 7th floor"+bcolors.ENDC)
column1 = Column(10, 2)
column1.elevatorList[0].status = "idle"
column1.elevatorList[0].position = 2
column1.elevatorList[0].direction = "none"
column1.elevatorList[1].status = "idle"
column1.elevatorList[1].position = 6
column1.elevatorList[1].direction = "none"
elevator = column1.requestElevator(3, "up")
column1.requestFloor(elevator, 7)

print(bcolors.HEADER + bcolors.UNDERLINE + bcolors.BOLD +
      "     [ Scenario 2 ]     "+bcolors.ENDC)
print(" 2.1. Someone is on the 1st floor and requests the 6th floor. ")
column1 = Column(10, 2)
column1.elevatorList[0].status = "idle"
column1.elevatorList[0].position = 10
column1.elevatorList[0].direction = "none"
column1.elevatorList[1].status = "idle"
column1.elevatorList[1].position = 3
column1.elevatorList[1].direction = "none"

elevator = column1.requestElevator(1, "up")
column1.requestFloor(elevator, 6)

print(bcolors.HEADER+"2.2. someone else is on the 3rd floor and requests the 5th floor. "+bcolors.ENDC)
elevator = column1.requestElevator(3, "up")
column1.requestFloor(elevator, 5)
print(bcolors.HEADER+"2.3. a third person is at floor 9 and wants to go down to the 2nd floor.  "+bcolors.ENDC)
elevator = column1.requestElevator(9, "down")
column1.requestFloor(elevator, 2)

print(bcolors.HEADER + bcolors.UNDERLINE + bcolors.BOLD +
      "     [ Scenario 3 ]     "+bcolors.ENDC)
print(bcolors.HEADER +
      " 3.1. Someone is on floor 3 and requests the 2nd floor.  "+bcolors.ENDC)
column1 = Column(10, 2)
column1.elevatorList[0].status = "idle"
column1.elevatorList[0].position = 10
column1.elevatorList[0].direction = "none"
column1.elevatorList[1].status = "moving"
column1.elevatorList[1].position = 3
column1.elevatorList[1].direction = "up"

elevator = column1.requestElevator(3, "down")
column1.requestFloor(elevator, 2)
print(bcolors.HEADER+"3.2. someone else is on the 10th floor and wants to go to the 3rd. "+bcolors.ENDC)


column1.elevatorList[0].status = "idle"
column1.elevatorList[0].position = 2
column1.elevatorList[0].direction = "none"
column1.elevatorList[1].status = "idle"
column1.elevatorList[1].position = 6
column1.elevatorList[1].direction = "none"

elevator = column1.requestElevator(10, "down")
column1.requestFloor(elevator, 3)
